### DISCUSSION TITLE
QNX compilation issues, vector.hpp Conflicts with version inherited from std vector

### DISCUSSION THREAD

**Discussion by rvishureddy** [duckdb/duckdb#8379] [Q&A]
Trying to compile duckdb to use on QNX and unable to compile with the qnx toolchain. 

vector.hpp:24:18: error: conflicts with version inherited from 'std::__1::vector<duckdb::LogicalType, std::__1::allocator<duckdb::LogicalType> >'

mkdir -p ./build/release && \
cd build/release && \
cmake -G "Ninja" -DFORCE_COLORED_OUTPUT=1        -DCMAKE_BUILD_TYPE=Release ../.. && \
cmake --build . --config Release
-- The C compiler identification is QCC 5.4.0
-- The CXX compiler identification is QCC 5.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/qnxtools/qnx700/host/linux/x86_64/usr/bin/qcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/qnxtools/qnx700/host/linux/x86_64/usr/bin/q++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found Git: /usr/bin/git (found version "2.17.1")
-- git hash 8d8f9f1b79, version v0.8.2-dev2090
CMake Warning at CMakeLists.txt:473 (message):
  Please use a recent compiler for debug builds

-- Load extension 'parquet' from '/home/vramire/duckdb-main/extensions'
-- Load extension 'jemalloc' from '/home/vramire/duckdb-main/extensions'
-- Extensions linked into DuckDB: parquet, jemalloc
-- Configuring done
-- Generating done
-- Build files have been written to: /home/vramire/duckdb-main/build/release
[1/394] Building CXX object src/planner/filter/CMakeFiles/duckdb_planner_filter.dir/ub_duckdb_planner_filter.cpp.o
FAILED: src/planner/filter/CMakeFiles/duckdb_planner_filter.dir/ub_duckdb_planner_filter.cpp.o
/usr/bin/ccache /home/qnxtools/qnx700/host/linux/x86_64/usr/bin/q++ -Vgcc_ntoaarch64le -lang-c++ -DDUCKDB -DDUCKDB_BUILD_LIBRARY -DDUCKDB_EXTENSION_JEMALLOC_LINKED=1 -DDUCKDB_EXTENSION_PARQUET_LINKED=1 -DDUCKDB_MAIN_LIBRARY -D_QNX_SOURCE -I/home/vramire/duckdb-main/src/include -I/home/vramire/duckdb-main/third_party/fsst -I/home/vramire/duckdb-main/third_party/fmt/include -I/home/vramire/duckdb-main/third_party/hyperloglog -I/home/vramire/duckdb-main/third_party/fastpforlib -I/home/vramire/duckdb-main/third_party/fast_float -I/home/vramire/duckdb-main/third_party/re2 -I/home/vramire/duckdb-main/third_party/miniz -I/home/vramire/duckdb-main/third_party/utf8proc/include -I/home/vramire/duckdb-main/third_party/miniparquet -I/home/vramire/duckdb-main/third_party/concurrentqueue -I/home/vramire/duckdb-main/third_party/pcg -I/home/vramire/duckdb-main/third_party/tdigest -I/home/vramire/duckdb-main/third_party/mbedtls/include -I/home/vramire/duckdb-main/third_party/jaro_winkler -I/home/vramire/duckdb-main/extension -I/home/vramire/duckdb-main/extension/parquet/include -I/home/vramire/duckdb-main/extension/jemalloc/include -O3 -DNDEBUG -O3 -DNDEBUG -O3 -DNDEBUG   -fPIC   -std=c++11 -fdiagnostics-color=always -std=c++11 -Wp,-MD,src/planner/filter/CMakeFiles/duckdb_planner_filter.dir/ub_duckdb_planner_filter.cpp.o.d -Wp,-MT,src/planner/filter/CMakeFiles/duckdb_planner_filter.dir/ub_duckdb_planner_filter.cpp.o -Wp,-MF,src/planner/filter/CMakeFiles/duckdb_planner_filter.dir/ub_duckdb_planner_filter.cpp.o.d -o src/planner/filter/CMakeFiles/duckdb_planner_filter.dir/ub_duckdb_planner_filter.cpp.o -c /home/vramire/duckdb-main/build/release/src/planner/filter/ub_duckdb_planner_filter.cpp
In file included from /home/vramire/duckdb-main/src/include/duckdb/common/common.hpp:13:0,
                 from /home/vramire/duckdb-main/src/include/duckdb/planner/table_filter.hpp:11,
                 from /home/vramire/duckdb-main/src/include/duckdb/planner/filter/conjunction_filter.hpp:11,
                 from /home/vramire/duckdb-main/src/planner/filter/conjunction_filter.cpp:1,
                 from /home/vramire/duckdb-main/build/release/src/planner/filter/ub_duckdb_planner_filter.cpp:2:
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp: In instantiation of 'class duckdb::vector<duckdb::LogicalType>':
/home/vramire/duckdb-main/src/include/duckdb/common/types.hpp:474:22:   required from here
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp:24:18: error: 'template<class _InputIterator> duckdb::vector<duckdb::LogicalType>::vector(_InputIterator, _InputIterator, const allocator_type&)' inherited from 'std::__1::vector<duckdb::LogicalType, std::__1::allocator<duckdb::LogicalType> >'
  using original::original;
                  ^
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp:24:18: error: conflicts with version inherited from 'std::__1::vector<duckdb::LogicalType, std::__1::allocator<duckdb::LogicalType> >'
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp: In instantiation of 'class duckdb::vector<duckdb::unique_ptr<duckdb::TableFilter> >':
/home/vramire/duckdb-main/src/include/duckdb/planner/filter/conjunction_filter.hpp:24:34:   required from here
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp:24:18: error: 'template<class _InputIterator> duckdb::vector<duckdb::unique_ptr<duckdb::TableFilter> >::vector(_InputIterator, _InputIterator, const allocator_type&)' inherited from 'std::__1::vector<duckdb::unique_ptr<duckdb::TableFilter>, std::__1::allocator<duckdb::unique_ptr<duckdb::TableFilter> > >'
/home/vramire/duckdb-main/src/include/duckdb/planner/filter/conjunction_filter.hpp:24:34:   required from here
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp:24:18: error: 'template<class _InputIterator> duckdb::vector<duckdb::unique_ptr<duckdb::TableFilter> >::vector(_InputIterator, _InputIterator, const allocator_type&)' inherited from 'std::__1::vector<duckdb::unique_ptr<duckdb::TableFilter>, std::__1::allocator<duckdb::unique_ptr<duckdb::TableFilter> > >'
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp:24:18: error: conflicts with version inherited from 'std::__1::vector<duckdb::unique_ptr<duckdb::TableFilter>, std::__1::allocator<duckdb::unique_ptr<duckdb::TableFilter> > >'
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp: In instantiation of 'class duckdb::vector<duckdb::ExceptionFormatValue>':
/home/vramire/duckdb-main/src/include/duckdb/parser/query_error_context.hpp:37:9:   required from here
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp:24:18: error: 'template<class _InputIterator> duckdb::vector<duckdb::ExceptionFormatValue>::vector(_InputIterator, _InputIterator, const allocator_type&)' inherited from 'std::__1::vector<duckdb::ExceptionFormatValue>'
/home/vramire/duckdb-main/src/include/duckdb/common/vector.hpp:24:18: error: conflicts with version inherited from 'std::__1::vector<duckdb::ExceptionFormatValue>'

**Comment by Tishj** [comment]
One is `std::vector`, the other is `duckdb::vector` not really sure why that's a problem.

The problem is likely in code where you include these headers, or the duckdb namespace together with the std namespace?

**Comment by rvishureddy** [comment]
I am just trying to compile the duckdb library. Not yet there in linking with my code. 

In vector.hpp
```template <class _Tp, bool SAFE = true>
class vector : public std::vector<_Tp, std::allocator<_Tp>> {
public:
        using original = std::vector<_Tp, std::allocator<_Tp>>;
        using original::original;
        using size_type = typename original::size_type;
        using const_reference = typename original::const_reference;
        using reference = typename original::reference;
```

> using original::original

is causing the issue.

**Reply by Tishj** [reply]
Please try to use this branch https://github.com/tishj/duckdb/tree/vector_explicit_constructor and see if that still causes a problem.

**Comment by rvishureddy** [comment]
@Tishj That didn't work either i still see the conflict between the duckdb vector and the standard vector with the same errors as described in the description.

**Reply by Tishj** [reply]
As no other system seems to have a problem with this I don't think I can help you further, the problem might just be with QCC and you need to make changes on your end to make this work.

The error says that there are conflicting names, but `duckdb::vector` and `std::vector` should be able to leave side by side peacefully.
Inside of `duckdb::vector` there is no explicit `using` of `std::vector` anymore in my branch, I was willing to buy that `using original::original` might produce colliding symbols but with that possibility eliminated I don't think we can fix this on our end.
